### PR TITLE
fix: prevent signup if username contains invalid characters

### LINKED ISSUE #2408: [feature request] prevent signup if username contains invalid characters
**Is your feature request related to a problem? Please describe.**
Username that is saved is different to the one entered by the user when joining as all special characters are stripped.  e.g. if you used sean@example.com as the username in the database it would save seanexamplecom

**Describe the solution you'd like**
I think it's preferable to tell users they can only use certain characters for the username and check this as part of the form validation prior to creating the user.

**Describe alternatives you've considered**
request a display name on the user signup form and then use an auto generated username based on the email address in the background.  Where the username currently appears in the ui then show the display name.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by jgable**
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

- Add a new `noSpecialCharacters` validator along with a `getSpecialCharacters` helper.
- Refactored the SignUp form to use the new validator via `composeValidators`.

## Git Issues

Closes #2408 

## Screenshots/Videos

Screenshot 2023-09-13 at 11 51 01 AM

I also tested the signup flow worked when using a display name without special characters and ran the SignUp Cypress tests locally.

### CODE REVIEW

File: src/utils/validators.test.ts [hunk 1]
```diff
@@ -64,3 +68,18 @@ describe('validateTitle', () => {
...
+  it('returns undefined when no special characters are present', () => {
+    const result = noSpecialCharacters('validUsername')
+
+    expect(result).toBeUndefined()
+  })
+
+  it('returns proper message for email values', () => {
+    const result = noSpecialCharacters('someones@email.com')
+
+    expect(result).toBeDefined()
```
Comment (AlfonsoGhislieri) [inline src/utils/validators.test.ts:82]: **Nitpick**: we don't need this assertion, we are already checking that the result will contain a string  after

File: src/utils/validators.ts [hunk 2]
```diff
@@ -13,6 +13,15 @@ import type { Mutator } from 'final-form'
 const required = (value: any) =>
   value ? undefined : 'Make sure this field is filled correctly'
 
+const noSpecialCharacters = (value: string) => {
+  const specialCharacters = getSpecialCharacters(value)
+  return specialCharacters.length > 0
+    ? `Cannot contain special characters; ${specialCharacters
+        .map((c) => `'${c}'`)
+        .join(', ')}`
+    : undefined
```
Comment (AlfonsoGhislieri) [inline src/utils/validators.ts:22]: **Issue**: not so sure about the error message here, it's a little confusing with the semi colon followed by a special characters in question..

Maybe we can use something simpler and easier for the user to understand at just a glance, eg: "Only letters and numbers are allowed." Taking this from [Microsoft 365](https://learn.microsoft.com/en-us/office/troubleshoot/office-suite-issues/username-contains-special-character)

**Review by AlfonsoGhislieri — CHANGES_REQUESTED** [review]
Thanks for this @jgable just a couple things I would change!

File: src/utils/validators.ts [hunk 3]
```diff
@@ -13,6 +13,15 @@ import type { Mutator } from 'final-form'
 const required = (value: any) =>
   value ? undefined : 'Make sure this field is filled correctly'
 
+const noSpecialCharacters = (value: string) => {
+  const specialCharacters = getSpecialCharacters(value)
+  return specialCharacters.length > 0
+    ? `Cannot contain special characters; ${specialCharacters
+        .map((c) => `'${c}'`)
+        .join(', ')}`
```
Comment (AlfonsoGhislieri) [inline src/utils/validators.ts:21]:
```suggestion
    ? "Only letters and numbers are allowed."
```

Issue: not so sure about the error message here, it's a little confusing with the semi colon followed by a special characters in question..

Maybe we can use something simpler and easier for the user to understand at just a glance, eg: "Only letters and numbers are allowed." Taking this from Microsoft 365

**Comment by AlfonsoGhislieri** [comment]
Thanks for the changes here @jgable, just linting test failing since prettier wasn’t run on one of the files (validator test file)

**Comment by AlfonsoGhislieri** [comment]
Thanks for the changes @jgable, we are failing the lint test in the `src/utils/validators.test.ts` file because prettier wasn't run on the file. Could we fix that please? :-)

https://prettier.io/

**Review by thisislawatts — APPROVED** [review]
Thanks @jgable for the time and effort on this PR ✨ Appreciate you for helping to build the community platform.

**Comment by thisislawatts** [comment]
@all-contributors add @jgable for code

**Pull request merged**